### DISCUSSION TITLE
How to build for iOS? What platform versions?

### DISCUSSION THREAD

**Discussion by AArnott** [AvaloniaUI/Avalonia#17203] [Q&A] [answered]
I can't find a winning combination of software and toolsets to target any iOS version.
I'm using .NET SDK 8.0.400.

### iOS 18

- Disqualifies most devices. 
- Requires Xcode 16, but Visual Studio can only work with Xcode 15.4. So I can only build this with `dotnet publish` at the CLI or in CI.

### iOS 17.5

- Only works with Visual Studio. The .NET SDK reports in `dotnet publish` that 17.5 isn't supported; only 17.0 and 18.0 are.

### iOS 17.2

Doesn't exist?

### iOS 17.0

- Build fails because Avalonia evidently targets 17.2, yet `net8.0-ios17.2` doesn't seem to be a real option.

> CSC : error CS1705: Assembly 'Avalonia.iOS' with identity 'Avalonia.iOS, Version=11.1.1.0, Culture=neutral, PublicKeyToken=c8d484a7012f9a8b' uses 'Microsoft.iOS, Version=17.2.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065' which has a higher version than referenced assembly 'Microsoft.iOS' with identity 'Microsoft.iOS, Version=17.0.0.0, Culture=neutral, PublicKeyToken=84e04ff9cfb79065' [/Users/buildagent/myagent/_work/1/s/src/Nerdbank.Zcash.App/Nerdbank.Zcash.App.iOS/Nerdbank.Zcash.App.iOS.csproj]

It looks like the Microsoft.iOS.dll that AvaloniaUI references comes from https://www.nuget.org/packages/Microsoft.iOS.Runtime.iossimulator-x64/17.2.8078. 
An arm64 equivalent exists too: https://www.nuget.org/packages/Microsoft.iOS.Runtime.iossimulator-arm64/17.2.8078

For some reason, my project.assets.json file reveals that it doesn't use either of these packages. It consumes Microsoft.iOS.Runtime.iossimulator-arm64.net8.0_17.0, for which 17.0 (not 17.2) is the latest available version.

Any suggestions?

**Comment by AArnott** [comment]
Changing my TargetFramework from `net8.0-ios17.0` to just `net8.0-ios` made that error go away. Not sure why. Hopefully it won't backfire in some other way. Getting all these things just right seems primarily luck and to be short-lived. :(

**Reply by maxkatz6** [reply]
For a user application it shouldn't backfire.
As `net8.0-ios` will target highest locally installed .NET iOS workload version. I.e. if current .NET iOS default is `ios18`, then it will be used.

It will backfire (like it did for us) for libraries. As it forces all library consumers to raise their target iOS version as well. Which is why we pinned `17.0`. But yes, for some reason SDK went with `Microsoft.iOS, Version=17.2.0.0`, I don't know why.

**Reply by maxkatz6** [reply]
Specifically, it backfired for us when minor .NET SDK update was raising this target version, essentially introducing unwanted breaking change breaking any .NET iOS application targeting older SDK.

**Comment by maxkatz6** [comment]
Avalonia doesn't target 17.2... It targets 17.0, making it compatible with anything higher too...
https://github.com/AvaloniaUI/Avalonia/blob/aabd038e88951a992124ddfcb2be06ee6daaf4a5/build/TargetFrameworks.props#L7

I really don't understand ways how .NET SDK behaves with these workloads.

**Comment by maxkatz6** [comment, marked as answer]
> 18.0 > Disqualifies most devices.

Note, that's not how it works.
TargetFramework version specifies a target version. I.e. iOS SDK version you want your app to be linked to.
This app still can run on older OSs assuming it doesn't use any newer APIs in runtime (version guards are useful there).

Actually, minimal support version is defined by SupportedOSPlatformVersion. And for Avalonia, supported version is 13.0.
https://github.com/AvaloniaUI/Avalonia/blob/master/src/iOS/Avalonia.iOS/Avalonia.iOS.csproj#L4C6-L4C32
https://github.com/AvaloniaUI/Avalonia/blob/11.2.0-rc1/build/TargetFrameworks.props#L20C6-L20C31

**Reply by AArnott** [reply]
It seems to me that `net8.0-ios18.0` sets both `TargetPlatformVersion` and `SupportedOSPlatformVersion` to 18.0.
Are you saying that I can use `net8.0-ios18.0` to use the latest SDK, but still set `SupportedOSPlatformVersion` to something older, like 17.0?

**Reply by maxkatz6** [reply]
> Are you saying that I can use net8.0-ios18.0 to use the latest SDK, but still set SupportedOSPlatformVersion to something older, like 17.0?

Correct.
